### PR TITLE
[ONNX][DORT] Lazy-import `onnxruntime`

### PR DESCRIPTION
**Pull request by oraluben**
Currently, if installed, `onnxruntime` will be imported when importing `torch._inductor` (which will be imported by some other library, e.g. transformer-engine):

```
  /mnt/c.py(53)<module>()
-> from torch._inductor.utils import maybe_profile
  /usr/local/lib/python3.10/site-packages/torch/_inductor/utils.py(49)<module>()
-> import torch._export
  /usr/local/lib/python3.10/site-packages/torch/_export/__init__.py(25)<module>()
-> import torch._dynamo
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/__init__.py(2)<module>()
-> from . import convert_frame, eval_frame, resume_execution
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py(48)<module>()
-> from . import config, exc, trace_rules
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/trace_rules.py(52)<module>()
-> from .variables import (
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/variables/__init__.py(38)<module>()
-> from .higher_order_ops import (
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/variables/higher_order_ops.py(14)<module>()
-> import torch.onnx.operators
  /usr/local/lib/python3.10/site-packages/torch/onnx/__init__.py(62)<module>()
-> from ._internal.onnxruntime import (
  /usr/local/lib/python3.10/site-packages/torch/onnx/_internal/onnxruntime.py(37)<module>()
-> import onnxruntime  # type: ignore[import]
```

This issue breaks generated triton kernel because it imported torch, and unexpected runtime libraries as well.

I've also added a test for this specific case under `test/onnx`, perhaps we should add more somewhere else?

Related issue: https://github.com/huggingface/accelerate/pull/3056

### CODE REVIEW

**Comment by justinchuby** [comment]
onnxruntime shouldn’t be imported at torch import because the whole torch.onnx module is lazily imported. Did I miss anything?

**Review by justinchuby — APPROVED** [review]

File: test/onnx/test_lazy_import.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+from torch.testing._internal import common_utils
+
+
+class TestLazyONNXRuntime(pytorch_test_common.ExportTestCase):
+
+    def test_lazy_import_(self):
+        import subprocess
+        import sys
+        import tempfile
```
Comment (justinchuby) [inline test/onnx/test_lazy_import.py:13]: Please move imports to the top of the file

File: test/onnx/test_lazy_import.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Owner(s): ["module: onnx"]
+
+import pytorch_test_common
+
+from torch.testing._internal import common_utils
+
+
+class TestLazyONNXRuntime(pytorch_test_common.ExportTestCase):
+
+    def test_lazy_import_(self):
```
Comment (justinchuby) [inline test/onnx/test_lazy_import.py:10]: test_onnxruntime_is_lazily_imported

**Comment by oraluben** [comment]
@justinchuby importing some submodule in torch might trigger importing `torch.onnx`, e.g.:

```
  /mnt/c.py(53)<module>()
-> from torch._inductor.utils import maybe_profile
  /usr/local/lib/python3.10/site-packages/torch/_inductor/utils.py(49)<module>()
-> import torch._export
  /usr/local/lib/python3.10/site-packages/torch/_export/__init__.py(25)<module>()
-> import torch._dynamo
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/__init__.py(2)<module>()
-> from . import convert_frame, eval_frame, resume_execution
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py(48)<module>()
-> from . import config, exc, trace_rules
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/trace_rules.py(52)<module>()
-> from .variables import (
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/variables/__init__.py(38)<module>()
-> from .higher_order_ops import (
  /usr/local/lib/python3.10/site-packages/torch/_dynamo/variables/higher_order_ops.py(14)<module>()
-> import torch.onnx.operators
  /usr/local/lib/python3.10/site-packages/torch/onnx/__init__.py(62)<module>()
-> from ._internal.onnxruntime import (
  /usr/local/lib/python3.10/site-packages/torch/onnx/_internal/onnxruntime.py(37)<module>()
-> import onnxruntime  # type: ignore[import]
```

**Comment by justinchuby** [comment]
I see. Thanks!

**Comment by justinchuby** [comment]
cc @xadupre @wschin

**Comment by justinchuby** [comment]
@oraluben Please run `lintrunner -a` to fix lint issues. After which this can be merged. Thanks!

File: test/onnx/test_lazy_import.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class TestLazyONNXRuntime(pytorch_test_common.ExportTestCase):
+
+    def test_onnxruntime_is_lazily_imported(self):
+
+        with tempfile.TemporaryDirectory() as wd:
+            r = subprocess.run([sys.executable, '-Ximporttime', '-c', 'import sys, torch.onnx'], capture_output=True, text=True, cwd=wd, check=True)
+
+        self.assertTrue(' onnxruntime' not in r.stderr, f'`onnxruntime` should not be imported, full importtime: {r.stderr}')
```
Comment (justinchuby) [inline test/onnx/test_lazy_import.py:19]:
```suggestion
        self.assertTrue('onnxruntime' not in r.stderr, f'`onnxruntime` should not be imported, full importtime: {r.stderr}')
```

Is the space intended?

Also this file needs formatting

Comment (oraluben) [reply test/onnx/test_lazy_import.py:19]: Yes that's a little bit hack but necessary. The space makes sure this test check for the `onnxruntime` library, not
`torch.onnx._internal.onnxruntime`.

Comment (justinchuby) [reply test/onnx/test_lazy_import.py:19]: In that case I would add a comment there

**Comment by oraluben** [comment]
unexpected imported `onnxscript` breaks some type annotation, I'm not sure about how the annotation part, trying to get more laziness to fix it.

**Comment by justinchuby** [comment]
Please do not alter the onnxscript imports. They should be fine

**Pull request closed without merging**